cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/DefaultSecureStoreService.java

```diff
@@ -0,0 +1,126 @@
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.SecureKeyId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.proto.security.SecureKeyCreateRequest;
+import co.cask.cdap.proto.security.SecureKeyListEntry;
+import co.cask.cdap.security.authorization.AuthorizerInstantiator;
+import co.cask.cdap.security.spi.authentication.SecurityRequestContext;
+import com.google.common.base.Strings;
+import com.google.inject.Inject;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Default implementation of the service that manages access to the Secure Store,
+ */
+public class DefaultSecureStoreService implements SecureStoreService {
+  private final AuthorizerInstantiator authorizerInstantiator;
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+  @Inject
+  DefaultSecureStoreService(AuthorizerInstantiator authorizerInstantiator, SecureStore secureStore,
+                            SecureStoreManager secureStoreManager) {
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+  }
+
+  /**
+   * Lists all the secure keys in the given namespace that the user has access to.
+   * @param namespaceId Id of the namespace we want the key list for.
+   * @return A list of {@link SecureKeyListEntry} of all the keys visible to the user under the given namespace.
+   * @throws Exception If either the caller does not have READ access to the namespace or if there was a problem
+   * getting the list from the underlying provider.
+   */
+  @Override
+  public List<SecureKeyListEntry> list(NamespaceId namespaceId) throws Exception {
+    Principal principal = SecurityRequestContext.toPrincipal();
+    authorizerInstantiator.get().enforce(namespaceId, principal, Action.READ);
+    List<SecureStoreMetadata> metadatas = secureStore.list(namespaceId.getNamespace());
+    List<SecureKeyListEntry> returnList = new ArrayList<>(metadatas.size());
+    for (SecureStoreMetadata metadata : metadatas) {
+      returnList.add(new SecureKeyListEntry(metadata.getName(), metadata.getDescription()));
+    }
+    return returnList;
+  }
+
+  /**
+   * Checks if the user has access to read the secure key and returns the data associated with the key if they do.
+   * @param secureKeyId Id of the key that the user is trying to read.
+   * @return Data associated with the key if the user has read access.
+   * @throws Exception If either the caller does not have READ access to the key or if there was a problem
+   * getting the data from the underlying provider.
+   */
+  @Override
+  public SecureStoreData get(SecureKeyId secureKeyId) throws Exception {
+    Principal principal = SecurityRequestContext.toPrincipal();
+    authorizerInstantiator.get().enforce(secureKeyId, principal, Action.READ);
```

Comment:
similarly, ``get`` should not enforce. It should use ``AuthorizationEnforcer.filter``. If that returns an empty list, then throw an ``UnauthorizedException``. See #6090 for details.